pkg/operator/status/status_controller.go

```diff
@@ -64,26 +62,24 @@ func NewClusterOperatorStatusController(
 	versionGetter VersionGetter,
 	recorder events.Recorder,
 ) *StatusSyncer {
-	c := &StatusSyncer{
+	return &StatusSyncer{
 		clusterOperatorName:   name,
 		relatedObjects:        relatedObjects,
 		versionGetter:         versionGetter,
 		clusterOperatorClient: clusterOperatorClient,
 		clusterOperatorLister: clusterOperatorInformer.Lister(),
 		operatorClient:        operatorClient,
-		eventRecorder:         recorder.WithComponentSuffix("status-controller"),
 		degradedInertia:       MustNewInertia(2 * time.Minute).Inertia,
-
-		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "StatusSyncer_"+strings.Replace(name, "-", "_", -1)),
+		controllerFactory: factory.New().ResyncEvery(time.Second).WithInformers(
+			operatorClient.Informer(),
+			clusterOperatorInformer.Informer(),
+		),
+		recorder: recorder.WithComponentSuffix("status-controller"),
 	}
+}
 
-	operatorClient.Informer().AddEventHandler(c.eventHandler())
-	clusterOperatorInformer.Informer().AddEventHandler(c.eventHandler())
-
-	c.cachesToSync = append(c.cachesToSync, operatorClient.Informer().HasSynced)
-	c.cachesToSync = append(c.cachesToSync, clusterOperatorInformer.Informer().HasSynced)
-
-	return c
+func (c *StatusSyncer) Run(ctx context.Context, workers int) {
+	c.controllerFactory.WithPostRunHooks(c.watchVersionGetterPostRunHook).WithSync(c.Sync).ToController("StatusSyncer_"+c.Name(), c.recorder).Run(ctx, workers)
```

Comment:
why not call all those With* funcs from the constructor. Continueing construction here in `Run` looks wrong.

Reply:
@sttts the reason is that we have mutating function that mutates the controller internal variables, which effectively changes the controller internal state... I saw that when you run Sync() then it might produce wrong results.

Reply:
I cannot follow. Please explain. This sounds very fishy.